File: compiler/src/pipeline/steps/typechecking.py
"""Type checking step implementation."""

from dataclasses import replace
from .base import MacroProcessingStep
from core.macro_registry import MacroContext, MacroRegistry
from core.node import Macro
from utils.logger import default_logger


class TypeCheckingStep(MacroProcessingStep):
    """Handles type checking"""
    
    def __init__(self, macros: MacroRegistry):
        super().__init__()
        # Use the unified typecheck registry
        self.macros = macros
        
    def process_node(self, ctx: MacroContext) -> None:
        """Type check a single node"""
        macro = str(ctx.compiler.get_metadata(ctx.node, Macro))
        all_macros = self.macros.all()
        
        # Create a description for this node for indentation
        node_desc = f"node {macro}"
        if hasattr(ctx.node, 'content') and ctx.node.content:
            # Limit content preview to keep it readable
            content_preview = ctx.node.content[:50] + ("..." if len(ctx.node.content) > 50 else "")
            node_desc = f"node {macro}: {content_preview}"
        
        with default_logger.indent("typecheck", node_desc):
            if macro in all_macros:
                with ctx.compiler.safely:
                    return all_macros[macro](ctx)
            else:
                for child in ctx.node.children:
                    child_ctx = replace(ctx, node=child)
                    self.process_node(child_ctx)

File: compiler/src_ts/pipeline/typechecking.ts
/**
 * Type checking step - TypeScript equivalent of pipeline/steps/typechecking.py
 */

import { MacroProcessingStep } from "./base.ts";
import { MacroContext, MacroRegistry } from "../core/macro_registry.ts";
import { Macrocosm } from "../core/macrocosm.ts";
import { defaultLogger } from "../utils/logger.ts";
import { ErrorType } from "../utils/error_types.ts";

export class TypeCheckingStep extends MacroProcessingStep {
	
	constructor(macros: MacroRegistry) {
		super(macros);
	}

	processNode(ctx: MacroContext): void {
		const compiler = ctx.compiler as Macrocosm;
		const macro = compiler.getMacro(ctx.node);
		const allMacros = this.macros.all();
		
		defaultLogger.typecheck(`typechecking macro: ${macro}`);
		
		// Look up macro provider
		const macroProvider = allMacros.get(macro.toString());
		if (!macroProvider) {
			compiler.compileError(
				ctx.node,
				`unknown macro: ${macro}`,
				ErrorType.INVALID_MACRO
			);
			return;
		}
		
		// Call the macro's typecheck method if it exists
		if (typeof macroProvider === 'object' && 'typecheck' in macroProvider) {
			const resultType = (macroProvider as { typecheck: (ctx: MacroContext) => string | null }).typecheck(ctx);
			// TODO: Store result type as metadata
			defaultLogger.typecheck(`macro ${macro} has type: ${resultType}`);
		}
		
		// Process children
		for (const child of ctx.node.children) {
			const childCtx: MacroContext = {
				...ctx,
				node: child,
			};
			this.processNode(childCtx);
		}
	}
}